chain/sync/heartbeat.go

```diff
@@ -0,0 +1,121 @@
+package sync
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"github.com/drand/drand/key"
+	"github.com/drand/drand/log"
+	"github.com/drand/drand/net"
+	"github.com/drand/drand/protobuf/drand"
+	clock "github.com/jonboulle/clockwork"
+)
+
+// HeartbeatConfig holds all the required information for the heartbeat to
+// proceed.
+type HeartbeatConfig struct {
+	// Logger where heartbeat will log important event such as network down
+	Log log.Logger
+	// Clock to regulate the heartbeats
+	Clock clock.Clock
+	// Client used by the heartbeat logic to contact other nodes
+	Client net.PublicClient
+	// Frequency at which the heartbeat should contact other nodes
+	Frequency time.Duration
+}
+
+// BeatInfo contains information relayed by the heartbeat logic. The peer who
+// replied and the last beacon that he has.
+type BeatInfo struct {
+	Peer      *key.Identity
+	LastRound uint64
+}
+
+// Heartbeat contacts other nodes in the network and relay their latest beacon
+type Heartbeat interface {
+	UpdateGroup([]*key.Identity)
+	Beats() chan BeatInfo
+	Stop()
+}
+
+type heartbeat struct {
+	c        *HeartbeatConfig
+	stop     chan bool
+	newGroup chan []*key.Identity
+	beats    chan BeatInfo
+}
+
+func NewHeartbeat(c *HeartbeatConfig) Heartbeat {
+	h := &heartbeat{
+		c:        c,
+		stop:     make(chan bool, 1),
+		newGroup: make(chan []*key.Identity, 1),
+		// TODO: It might work for now but wont for bigger size
+		beats: make(chan BeatInfo, 100),
+	}
+	go h.run()
+	return h
+}
+
+func (h *heartbeat) Stop() {
+	close(h.stop)
+}
+
+func (h *heartbeat) Beats() chan BeatInfo {
+	return h.beats
+}
+
+func (h *heartbeat) UpdateGroup(g []*key.Identity) {
+	h.newGroup <- g
+}
+
+func (h *heartbeat) run() {
+	var currentGroup []*key.Identity
+	ctx, cancel := context.WithCancel(context.Background())
+	for {
+		select {
+		case ng := <-h.newGroup:
+			if ng == nil {
+				h.c.Log.Error("heartbeat", "received nil group")
+				continue
+			}
+			currentGroup = ng
+			cancel() // we cancel current sync operations
+			ctx, cancel = context.WithCancel(context.Background())
+			go h.fetch(ctx, currentGroup)
```

Comment:
this should be under the clock tick, not under the newgroup select, i think

Reply:
Right - I ended up doing an aggregated solution with the max value directly computed and so only one value is emitted at each heartbeat.